sql: internal error: while building cascade expression: runtime error: index out of range [2] with length 2

<hidden>

roachtest.sqlsmith/setup=rand-tables/setting=no-ddl [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2921792&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2921792&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-ddl) on master @ [2d7f9a08bac49040a6d77c5465e2b4fd8fdac1f4](https://github.com/cockroachdb/cockroach/commits/2d7f9a08bac49040a6d77c5465e2b4fd8fdac1f4):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-ddl/run_1
	sqlsmith.go:226,sqlsmith.go:262,test_runner.go:777: error: pq: internal error: while building cascade expression: runtime error: index out of range [2] with length 2
		stmt:
		DELETE FROM
			defaultdb.public.table2 AS tab_8870
		WHERE
			EXISTS(
				WITH
					with_2844 (col_19876, col_19877)
						AS (SELECT * FROM (VALUES (NULL, B'1101100000111')) AS tab_8871 (col_19876, col_19877))
				SELECT
					tab_8870.col2_6 AS col_19878
				FROM
					with_2844 AS cte_ref_828, defaultdb.public.table4@[0] AS tab_8872
				WHERE
					true
				LIMIT
					4:::INT8
			);
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh sqlsmith/setup=rand-tables/setting=no-ddl
```
</p>
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-ddl.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/cascade
+++ b/pkg/sql/logictest/testdata/logic_test/cascade
@@ -4352,3 +4352,12 @@ SELECT * FROM child
 
 statement ok
 DROP TABLE child, parent
+
+# Regression test for #64179. A cascading delete with a subquery should not
+# error.
+statement ok
+CREATE TABLE a (a INT UNIQUE);
+CREATE TABLE b (b INT, FOREIGN KEY (b) REFERENCES a (a) ON DELETE CASCADE);
+
+statement ok
+DELETE FROM a WHERE EXISTS (SELECT a FROM a)

--- a/pkg/sql/opt/optbuilder/fk_cascade.go
+++ b/pkg/sql/opt/optbuilder/fk_cascade.go
@@ -189,6 +189,9 @@ func tryNewOnDeleteFastCascadeBuilder(
 		if memo.CanBeCompositeSensitive(md, &sel.Filters) {
 			return nil, false
 		}
+		if sel.Relational().HasSubquery {
+			return nil, false
+		}
 		filters = sel.Filters
 
 	case opt.ScanOp:

--- a/pkg/sql/opt/optbuilder/testdata/fk-on-delete-cascade
+++ b/pkg/sql/opt/optbuilder/testdata/fk-on-delete-cascade
@@ -65,6 +65,42 @@ root
                 └── filters
                      └── child.p:9 = p:11
 
+# Delete with subquery; no fast path.
+build-cascades
+DELETE FROM parent WHERE EXISTS (SELECT p FROM parent)
+----
+root
+ ├── delete parent
+ │    ├── columns: <none>
+ │    ├── fetch columns: p:3
+ │    ├── input binding: &1
+ │    ├── cascades
+ │    │    └── fk_p_ref_parent
+ │    └── select
+ │         ├── columns: p:3!null crdb_internal_mvcc_timestamp:4
+ │         ├── scan parent
+ │         │    └── columns: p:3!null crdb_internal_mvcc_timestamp:4
+ │         └── filters
+ │              └── exists
+ │                   └── project
+ │                        ├── columns: p:5!null
+ │                        └── scan parent
+ │                             └── columns: p:5!null crdb_internal_mvcc_timestamp:6
+ └── cascade
+      └── delete child
+           ├── columns: <none>
+           ├── fetch columns: c:10 child.p:11
+           └── semi-join (hash)
+                ├── columns: c:10!null child.p:11!null
+                ├── scan child
+                │    └── columns: c:10!null child.p:11!null
+                ├── with-scan &1
+                │    ├── columns: p:13!null
+                │    └── mapping:
+                │         └──  parent.p:3 => p:13
+                └── filters
+                     └── child.p:11 = p:13
+
 # Delete everything.
 build-cascades
 DELETE FROM parent

